### PR TITLE
bcache: avoid deadlock between makeRoomForSize and onEvict

### PR DESCRIPTION
**Pull request by strib**
The deadlock:

* golang-lru takes its internal lock, calls b.onEvict, which takes bytesLock
* b.makeRootForSize takes bytesLock, calls cleanTransient.Len(), which takes
  the golang-lru internal lock.

Instead, only access cleanTransient.Len() outside of the bytesLock
critical section.

Issue: KBFS-1336

### CODE REVIEW

**Comment by akalin-keybase** [comment]
Hmm...I think the deeper problem is that we have too many locks than we need in bcache, and that a better solution would be to get rid of some of them.

Specifically, since we only evict at known places (`makeRoomForSize` and `deleteTransient`) we don't really need the `onEvict` callback, and we can just do it inline. It's less DRY since we have to remember to decrement `cleanTotalBytes`, but empirically it looks like using `onEvict` is more problematic!

The problem is that `RemoveOldest` doesn't return the removed entry. However, the non-threadsafe LRU version (`simplelru.LRU`) does, so we can just use that, and make `cleanLock` protect `cleanTransient` too, which makes more sense anyway. Then the whole last half of `Put` can just be under `cleanLock`.

Also, I don't think we need a separate lock for `cleanTotalBytes`, either, since we only read/write it when messing with `cleanTransient` or `cleanPermanent`.

So to summarize, I propose:

1) Making `cleanTransient` a `simplelru.LRU` and protecting it with `cleanLock`, and do the size calculations manually instead of using `onEvict`.
2) Removing `bytesLock` and protecting `cleanTotalBytes` with `cleanLock`.

What do you think?

**Comment by strib** [comment]
I don't think that works, because other `Put`s can still evict items in the cache (if it hits the max size limit), and we wouldn't be able to track those bytes without `onEvict`.

**Comment by akalin-keybase** [comment]
What other `Put`s can evict stuff from `cleanTransient`? The only place I see is the `cleanTransient.Add` call in `Put`, which is done immediately after manually evicting in `makeRoomForSize`.

**Comment by strib** [comment]
But `makeRoomForSize` only evicts if we're over the bytes limit.  There's a "number of entries" built into `golang-lru` and `Add()` will evict based on that.

**Comment by strib** [comment]
The real solution is to build our own LRU that only evicts based on bytes size.  But that's way out of scope for this fix.

**Comment by akalin-keybase** [comment]
Ah, okay, I get it now. Let me take another look...

**Comment by akalin-keybase** [comment]
Okay, how about this:

1) Remove `bytesLock` and protect `cleanTotalBytes` with `cleanLock`.
2) Make `cleanTransient` a `simplelru.LRU` and protect it with `cleanLock`. Keep the current `onEvict` behavior, and ensure/document that `cleanLock` is always held on any operation that could evict, so `onEvict` doesn't take any locks.

**Comment by strib** [comment]
`cleanLock` isn't held on all operations that can evict.  ( The `Add` call in `Put` specifically)  Is that step 1.5?  I guess that would work but I'm a little nervous about it since we've never tried `simpleLRU` and I wanted this to be a simple change.

**Comment by akalin-keybase** [comment]
Okay. The current code LGTM, so maybe put a TODO repeating/linking to the discussion in this issue?
